---
title: "01.1 MetaData"
output: html_document
date: "2025-21-01"
params: 
  chIPFilesDir: 
    label: "chIPFilesDir"
    value: ./chIPdata
  atacFilesDir:
    label: "atacFilesDir"
    value: ./atacData
  outDir: 
    label: "outDir"
    value: ./outDir
---

```{r setup}
library(data.table)
library(stringr)
```

```{r, paths}
chIPFilesDirBase <- params$chIPFilesDir
atacFilesDir <- params$atacFilesDir

chIPFilesDir1 <- file.path(chIPFilesDirBase, "GTRD")
chIPFilesDir2 <- file.path(chIPFilesDirBase, "ENCODE")
chIPFilesDir3 <- file.path(chIPFilesDirBase, "codebook")
chIPFilesDir4 <- file.path(chIPFilesDirBase, "brainTF") 

metaEncodePath <- "../data/meta_data/fileListEncode.tsv"
#metaEncodePath <- "../data/meta_data/chip_metadata.rds"

outDir <- params$outDir

absPath <- function(baseDir, relativePath){
  dir <- file.path(baseDir, relativePath)
  return(dir)
}
```

# GTRD metadata

## Download meta data
```{r, meta GTRD: download}
metaAtacGtrd <- fread("http://gtrd.biouml.org:8888/downloads/20.06/metadata/ATAC-seq.metadata.txt")
metaChIPGtrd <- fread("http://gtrd.biouml.org:8888/downloads/20.06/metadata/ChIP-seq.metadata.txt")
metaExt <- fread("http://gtrd20-06.biouml.org/downloads/20.06/metadata/external_references.metadata.txt")
```

Get cell state ids (cellular context x treatment)
```{r, meta GTRD: subset to species}
metaAtacGtrd <- subset(metaAtacGtrd, organism=="Homo sapiens")
metaChIPGtrd <- subset(metaChIPGtrd, specie=="Homo sapiens")
```

## Unify treatment names

```{r, meta GTRD: control / wildtype / no treatment labels}
controlLabels <- c("Genotype: wildtype;",
                   "Genotype: wildtype; Treatment: none",
                   "Genotype: control (wildtype);",
                   "Treatment: no Genotype: wildtype",
                   "Grnotype: wildtype", "none",
                   "Treatment: Genotype: wildtype",
                   "Treatment: none; Genotype: wildtype",
                   "Treatment: Treatment: none; Genotype: wildtype; Genotype: wildtype",
                   "Genotype: wildtype(control);", "none", 
                   "Genotype: control;",
                   "control;",
                   "none (no stimulation control)",
                   "Treatment: none (Control);",
                   "Treatment: Control (none);",
                   "Treatment: control (none);",
                   "treatment: none",
                   "Treatment: Treatment: none",
                   "Treatment: none; Conditions: control",
                   "Treament: none",
                   "Genotype: wildtype; Treatment: none;",
                   "Treatment: none; Time point: 0 hours",
                   "Treatment: control",
                   "Treatment: untreated, control",
                   "Treatment: Treatment: control",
                   "Treatment: none",
                   "Treatment: none;",
                   "",
                   " ",
                   "none")
```

Unifying control/wildtype/no treatments
```{r, meta GTRD: unifying nones / wildtypes / controls}
metaAtacGtrd[,treatment:=fifelse(treatment %in% controlLabels,
                                 "none", treatment)]
metaChIPGtrd[,treatment:=fifelse(treatment %in% controlLabels,
                                 "none", treatment)]
```

## External Ids

IDs of other databases
```{r, meta GTRD: external DB IDs}
metaExtGEOSample <- subset(metaExt, external_db=="GEO" & 
                           grepl("GSM", external_db_id))
metaExtGEOExp <- subset(metaExt, external_db=="GEO" &
                        grepl("GSE", external_db_id))
metaExtOther <- subset(metaExt, external_db!="GEO")

metaExtGEOSample <- dcast(metaExtGEOSample, id~external_db, 
                          value.var="external_db_id", 
                          fun.aggregate=paste, collapse=";", 
                          fill=NA)
setnames(metaExtGEOSample, "GEO","GEO_sample")

metaExtGEOExp <- dcast(metaExtGEOExp, id~external_db, 
                          value.var="external_db_id", 
                          fun.aggregate=paste, collapse=";", 
                          fill=NA)
setnames(metaExtGEOExp, "GEO","GEO_experiment")

metaExtOther <- dcast(metaExtOther, id~external_db, 
                      value.var="external_db_id", 
                      fun.aggregate=paste, collapse=";", 
                      fill=NA)

metaExt <- merge(metaExtGEOSample, metaExtGEOExp, 
                 by.x=c("id"), by.y=c("id"), all=TRUE)
metaExt <- merge(metaExt, metaExtOther, 
                  by.x=c("id"), by.y=c("id"), all=TRUE)
```

## Annotate and subset metadata

```{r, meta GTRD: subset ATAC meta}
atacFilePaths <- list.files(atacFilesDir, full.names=TRUE, recursive=TRUE)
atacFilePaths <- atacFilePaths[grepl("AEXP", atacFilePaths) & 
                               grepl("fragsInDhs.rds", atacFilePaths)]

atacFilePaths <- data.table(file_path=atacFilePaths)
atacFilePaths <- subset(atacFilePaths, file.exists(atacFilePaths$file_path))
atacFilePaths[,file_name:=basename(dirname(file_path))]
atacFilePaths[,c("geo_id", "gtrd_id"):=tstrsplit(file_name, 
                                                 split=".", fixed=TRUE)]
metaAtacGtrd <- merge(atacFilePaths, 
                      metaAtacGtrd, 
                      by.x=c("gtrd_id"), by.y=c("id"), all.x=TRUE, all.y=FALSE)

setnames(metaAtacGtrd, "geo_id", "external_id_atac")
setnames(metaAtacGtrd, "file_path", "atac_path")
setnames(metaAtacGtrd, "gtrd_id", "gtrd_atac_id")
setnames(metaAtacGtrd, "file_name", "file_name_atac")
```

```{r, meta GTRD: annotate ATAC}
metaAtacGtrd <- merge(metaAtacGtrd, metaExt, 
                      by.x=c("gtrd_atac_id"), 
                      by.y=c("id"))
extDbs <- c("ENCODE", "ENCODEPROJ", "GEO_sample", 
            "GEO_experiment", "PUBMED", "SRA")
setnames(metaAtacGtrd, extDbs,
         paste(extDbs, "ATAC", sep="_"))

metaAtacGtrd <-  metaAtacGtrd[,c("gtrd_atac_id", "file_name_atac", "title",
                                 "treatment", "atac_path", 
                                 paste(extDbs, "ATAC", sep="_")), 
                         with=FALSE]
length(unique(metaAtacGtrd$input))
length(unique(metaAtacGtrd$gtrd_atac_id))

# to later on get a global id
metaAtacGtrd[,db_id:=gtrd_atac_id]
metaAtacGtrd$origin <- "GTRD"

saveRDS(metaAtacGtrd, absPath(outDir, "annotatedMetaGTRDATAC.rds"))
```

```{r, meta GTRD: subset ChIP meta}
peaksMACS2 <- list.files(absPath(chIPFilesDir1, 
                                 "MACS2_matched_ids/MACS2_chIP_bed"))
peaksPICS2 <- list.files(absPath(chIPFilesDir1, 
                                 "PICS_matched_ids/PICS_chIP_bed"))
peaksSISSR <- list.files(absPath(chIPFilesDir1, 
                                 "SISSRS_matched_ids/SISSRS_chIP_bed"))
peaksGEM <- list.files(absPath(chIPFilesDir1, 
                               "GEM_matched_ids/GEM_chIP_bed"))

pathsMACS2 <- list.files(absPath(chIPFilesDir1, 
                                 "MACS2_matched_ids/MACS2_chIP_bed"), 
                         full.names=TRUE)
pathsPICS2 <- list.files(absPath(chIPFilesDir1, 
                                 "PICS_matched_ids/PICS_chIP_bed"), 
                         full.names=TRUE)
pathsSISSR <- list.files(absPath(chIPFilesDir1, 
                                 "SISSRS_matched_ids/SISSRS_chIP_bed"), 
                         full.names=TRUE)
pathsGEM <- list.files(absPath(chIPFilesDir1, 
                               "GEM_matched_ids/GEM_chIP_bed"), 
                       full.names=TRUE, recursive=TRUE)

peakInfo <- data.table(file_name=c(peaksMACS2, peaksPICS2, 
                                   peaksSISSR, peaksGEM),
                       chip_paths=c(pathsMACS2, pathsPICS2, 
                                    pathsSISSR, pathsGEM))
peakInfo[,c("peak_id", "tf_name", "uniprot_id", 
            "caller", "number"):=tstrsplit(file_name, split="_")]

metaChIPGtrd <- merge(peakInfo, metaChIPGtrd, 
                  all.x=TRUE, all.y=FALSE,
                  by.x="peak_id", 
                  by.y=c("input"))
```

```{r, meta GTRD: annotate ChIP data}
metaChIPGtrd$is_chip <- TRUE
metaChIPGtrd <- merge(metaChIPGtrd, metaExt,
                      by.x=c("id"), by.y=c("id"), all.x=TRUE, all.y=FALSE)

setnames(metaChIPGtrd, "id", "gtrd_chip_id")
setnames(metaChIPGtrd, "file_name", "file_name_chip")
setnames(metaChIPGtrd, extDbs,
         paste(extDbs, "chIP", sep="_"))

metaChIPGtrd <- metaChIPGtrd[,c("gtrd_chip_id", "peak_id", "file_name_chip",
                                "chip_paths", "tf_name", "uniprot_id", 
                                "treatment", "title", "caller",
                           paste(extDbs, "chIP", sep="_")), with=FALSE]
length(unique(metaChIPGtrd$peak_id))
length(unique(metaChIPGtrd$gtrd_chip_id))
metaChIPGtrd$origin <- "GTRD"

# to later on get a global id
metaChIPGtrd[,db_id:=peak_id]
saveRDS(metaChIPGtrd, absPath(outDir, "annotatedMetaGTRDChIP.rds"))
```

# ENCODE metadata

```{r, meta ENCODE: meta ChIP}
chIPEncodeFiles <- list.files(chIPFilesDir2, 
                              full.names=TRUE, recursive=TRUE)
chIPEncodeFileNames <- list.files(chIPFilesDir2, recursive=TRUE)
chIPEncodeFiles <- data.table(chip_paths_encode=chIPEncodeFiles, 
                              base_dir=dirname(chIPEncodeFileNames),
                              file_names=basename(chIPEncodeFileNames))
chIPEncodeFiles[,c("encodeFileID", "tf_name"):=tstrsplit(file_names, 
                                                     split="_", keep=1:2)]
chIPEncodeFiles[,context:=base_dir]

metaChIPEncode <- fread(metaEncodePath)
metaChIPEncode$from_encode <- TRUE
metaChIPEncode[,encodeFileID:=Accession]
metaChIPEncode[,encodeDatasetID:=basename(Dataset)]
setnames(metaChIPEncode, "Target label", "tf_targets")
metaChIPEncode <- merge(metaChIPEncode[,c("encodeFileID", "encodeDatasetID",
                                          "tf_targets")], 
                        chIPEncodeFiles, 
                        by.x=c("encodeFileID"), by.y=c("encodeFileID"))
                        #by.x=c("tf_name", "context"),
                        #by.y=c("tf_name", "context"), allow.cartesian=TRUE)
metaChIPEncode$origin <- "ENCODE"
setnames(metaChIPEncode, "context", "title")
metaChIPEncode$treatment <- NA
metaChIPEncode$caller <- "MACS2"
setnames(metaChIPEncode, "chip_paths_encode", "chip_paths")
metaChIPEncode[,db_id:=encodeFileID]
```

```{r, meta ENCODE: meta ATAC}
atacDirs <- list.dirs(atacFilesDir, recursive=TRUE)
atacDirNames <- basename(atacDirs)
toKeep <- grepl("encode", atacDirs)
atacDirs <- atacDirs[toKeep]
atacDirNames <- atacDirNames[toKeep]
metaAtacEncode <- data.table(atac_path=file.path(atacDirs, "fragsInDhs.rds"), 
                             title=atacDirNames)

metaAtacEncode <- subset(metaAtacEncode, !grepl("see_origin", title) & 
                                         !grepl("forBrainTF", title) & 
                                         !grepl("trimmed", title) & 
                           title!="" & title!="merge" & 
                           file.exists(atac_path))
metaAtacEncode[,title:=fifelse(grepl("HEK293", title), "HEK293", title)]
metaAtacEncode$origin <- "ENCODE"
metaAtacEncode$db_id <- NA
metaAtacEncode$treatment <- NA
```

# Codebook metadata

```{r, meta Codebook: meta ChIP}
cbFilesNames <- list.files(chIPFilesDir3)
cbFiles <- list.files(chIPFilesDir3, full.names=TRUE)
metaChIPCb <- data.table(file_names=cbFilesNames,
                      chip_paths=cbFiles)
metaChIPCb[,c("db_id","tf_name", "type"):=tstrsplit(file_names, 
                                                    split="_", keep=c(1:3))]
metaChIPCb <- subset(metaChIPCb, grepl("ChIP", type))
metaChIPCb$title <- "HEK293"
metaChIPCb$treatment <- "none"
metaChIPCb$origin <- "codebook"
metaChIPCb$caller <- "MACS2"
```

# BrainTF metadata

```{r, meta BrainTF: meta ChIP}
brainTfFiles <- list.files(chIPFilesDir4, full.names=TRUE)
brainTfFileNames <- list.files(chIPFilesDir4)

metaChIPBrainTf <- data.table(file_names=brainTfFileNames,
                              chip_paths=brainTfFiles)
metaChIPBrainTf[,c("context", "tf_name"):=tstrsplit(file_names,
                                                    split="_", keep=1:2)]
metaChIPBrainTf[,title:=gsub(tf_name, "", context),
                 by=seq_len(nrow(metaChIPBrainTf))]

metaChIPBrainTf[,title_broad:=fifelse(grepl("neurons", title), 
                                      "neurons",title)]
metaChIPBrainTf[,title_broad:=fifelse(grepl("oligos", title_broad), 
                                      "oligos",title_broad)]

metaChIPBrainTf <- subset(metaChIPBrainTf, 
                          title_broad %in% c("neurons", "oligos"))
metaChIPBrainTf[,title:=title_broad]
metaChIPBrainTf$title_broad <- NULL

metaChIPBrainTf$origin <- "brainTF"
metaChIPBrainTf$treatment <- "none"
metaChIPBrainTf$caller <- "MACS2"

metaChIPBrainTf[,db_id:=context] # unclean
```

```{r, meta BrainTF: meta ATAC}
atacDirs <- list.dirs(atacFilesDir, recursive=TRUE, full.names=TRUE)
atacDirNames <- basename(atacDirs)
toKeep <- grepl("forBrainTF", atacDirs)

atacDirs <- atacDirs[toKeep]
atacDirNames <- atacDirNames[toKeep]
metaAtacBrainTf <- data.table(atac_path=file.path(atacDirs, 
                                                 "fragsInDhs.rds"), 
                              title=atacDirNames)

metaAtacBrainTf <- subset(metaAtacBrainTf, grepl("forBrainTF", title) & 
                                          !grepl("see_origin", title) & 
                            title!="" & title!="merge" & 
                            file.exists(atac_path))
metaAtacBrainTf[,title:=gsub("forBrainTF.", "", title)]
metaAtacBrainTf$treatment <- "none"
metaAtacBrainTf[,db_id:=title]
metaAtacBrainTf$origin <- "brainTF"
```

# Other sources

ATAC separately processed from ENCODE/GEO
```{r, meta manual: meta ATAC}
atacDirs <- list.dirs(atacFilesDir, recursive=TRUE, full.names=TRUE)
atacDirNames <- basename(atacDirs)

toKeep <- grepl("manual", atacDirs)
atacDirs1 <- atacDirs[toKeep]
atacDirNames1 <- atacDirNames[toKeep]

toKeep <- grepl("prediction", atacDirs) 
atacDirs2 <- atacDirs[toKeep]
atacDirNames2 <- atacDirNames[toKeep]

# remove contexts which are present in other files
toKeep2 <- !(atacDirNames2 %in% atacDirNames1) & 
           !(atacDirNames2 %in% metaAtacEncode$title)
atacDirs2 <- atacDirs2[toKeep2]
atacDirNames2 <- atacDirNames2[toKeep2]

atacDirs <- c(atacDirs1, atacDirs2)
atacDirNames <- c(atacDirNames1, atacDirNames2)

metaAtacManual <- data.table(atac_path=file.path(atacDirs, "fragsInDhs.rds"), 
                             title=atacDirNames)
metaAtacManual <- subset(metaAtacManual, !grepl("trimmed", title) & 
                                         !grepl("SRX", title) & 
                                         !grepl("forBrainTF", title) &
                           !grepl("aligned", title) &
                           file.exists(atac_path))
metaAtacManual$treatment <- "none"
metaAtacManual[,db_id:=title]
metaAtacManual$origin <- "manual"
```

# Merge and match metadata

## Match GTRD

```{r, match ChIP & ATAC GTRD}
atacCols <- c("db_id", "title", "treatment", "atac_path", "origin")
metaMatchGTRD <- merge(metaChIPGtrd,
                       rbind(metaAtacGtrd[,atacCols,with=FALSE], 
                             metaAtacManual[,atacCols,with=FALSE]),
                       by.x=c("title", "treatment"), 
                       by.y=c("title", "treatment"),
                       all.x=TRUE,
                       all.y=FALSE, allow.cartesian=TRUE)
setnames(metaMatchGTRD, c("db_id.x", "db_id.y"), 
                        c("db_id_chIP", "db_id_ATAC"))
setnames(metaMatchGTRD, c("origin.x", "origin.y"), 
                        c("origin_chIP", "origin_ATAC"))
```

## Match ENCODE

```{r, match ChIP & ATAC ENCODE}
atacCols <- c("db_id", "title", "atac_path", "origin")
metaMatchEncode <- merge(metaChIPEncode, 
                         rbind(metaAtacEncode[,atacCols,with=FALSE],
                               metaAtacManual[,atacCols,with=FALSE]),
                         by.x=c("title"), by.y=c("title"),
                         all=TRUE, allow.cartesian=TRUE)
setnames(metaMatchEncode, c("db_id.x", "db_id.y"), 
                          c("db_id_chIP", "db_id_ATAC"))
setnames(metaMatchEncode, c("origin.x", "origin.y"), 
                          c("origin_chIP", "origin_ATAC"))
```

## Match brainTf

```{r}
atacCols <- c("db_id", "title", "treatment", "atac_path", "origin")
metaMatchBrainTf <- merge(metaChIPBrainTf, 
                          metaAtacBrainTf[,atacCols,with=FALSE],
                          by.x=c("title", "treatment"), 
                          by.y=c("title", "treatment"),
                          all=TRUE, allow.cartesian=TRUE)
setnames(metaMatchBrainTf, c("db_id.x", "db_id.y"), 
                           c("db_id_chIP", "db_id_ATAC"))
setnames(metaMatchBrainTf, c("origin.x", "origin.y"), 
                           c("origin_chIP", "origin_ATAC"))
```

## Match Codebook

```{r}
atacCols <- c("db_id", "title", "treatment", "atac_path", "origin")
metaMatchCb <- merge(metaChIPCb, metaAtacManual, 
                     by.x=c("title","treatment"), 
                     by.y=c("title","treatment"), 
                     all=TRUE,
                     allow.cartesian=TRUE)
setnames(metaMatchCb, c("db_id.x", "db_id.y"), 
                      c("db_id_chIP", "db_id_ATAC"))
setnames(metaMatchCb, c("origin.x", "origin.y"), 
                      c("origin_chIP", "origin_ATAC"))
```

## Merge meta-data

```{r, merge metadata}
# prioritize ENCODE data if both is there
metaMatchGTRDSub <- subset(metaMatchGTRD, 
                           !(metaMatchGTRD$ENCODEPROJ_chIP %in% metaMatchEncode$encodeDatasetID))
# 
# atacCols <- c("db_id_ATAC", "title", "treatment", "atac_path", "origin_ATAC")
# metaAtacAll <- rbind(metaMatchGTRDSub[,atacCols,with=FALSE],
#                      metaMatchEncode[,atacCols,with=FALSE],
#                      metaMatchBrainTf[,atacCols,with=FALSE],
#                      metaMatchCb[,atacCols,with=FALSE])
# metaAtacAll <- unique(metaAtacAll, by="atac_path")
# 
# chIPCols <- c("db_id_chIP", "title", "treatment", "chip_paths", "origin_chIP", "tf_name")
# metaChIPcAll <- rbind(metaMatchGTRDSub[,chIPCols,with=FALSE],
#                      metaMatchEncode[,chIPCols,with=FALSE],
#                      metaMatchBrainTf[,chIPCols,with=FALSE],
#                      metaMatchCb[,chIPCols,with=FALSE])
# metaChIPAll <- unique(metaChIPcAll, by="chip_paths")
# 
# metaChIPAll[,treatment:=fifelse(is.na(treatment), "NA", treatment)]
# metaAtacAll[,treatment:=fifelse(is.na(treatment), "NA", treatment)]
# metaMatched <- merge(metaChIPAll, metaAtacAll, 
#                      by.x=c("title", "treatment"), 
#                      by.y=c("title", "treatment"),
#                      all=TRUE, allow.cartesian=TRUE)
# metaMatched[,is_matched:=fifelse(!is.na(atac_path) & !is.na(chip_paths), TRUE, FALSE)]
# length(unique(subset(metaMatched, is_matched)$tf_name))

# alternative
metaMatch <- rbindlist(list(metaMatchGTRDSub, 
                            metaMatchEncode, 
                            metaMatchCb, 
                            metaMatchBrainTf), fill=TRUE, use.names=TRUE)
metaMatch[,is_matched:=fifelse(!is.na(atac_path) & !is.na(chip_paths), TRUE, FALSE)]
length(unique(subset(metaMatch, is_matched)$tf_name))
length(unique(subset(metaMatch, is_matched)$title))
length(unique(subset(metaMatch, is_matched)$treatment))
metaMatch[,context:=paste(title, treatment,sep="_")]
metaMatch[,context:=gsub("_NA", "", context)]
length(unique(subset(metaMatch, is_matched)$context))

metaMatch[,context_id:=as.integer(factor(context,
                                         levels=unique(metaMatch$context)))]
metaMatch[,tf_id:=as.integer(factor(tf_name, levels=unique(metaMatch$tf_name)))]
metaMatch[,combination_id:=paste(tf_name, context, sep="_")]
metaMatch[,combination_id:=as.integer(factor(combination_id, 
                                      levels=unique(metaMatch$combination_id)))]

metaMatchAll <- metaMatch
metaMatch <- subset(metaMatch, is_matched)

saveRDS(metaMatch, absPath(outDir, "01_fullMetaMatched.rds"))
saveRDS(metaMatchAll, absPath(outDir, "01_fullMetaAll.rds"))
```

```{r, sessioninfo}
sessionInfo()
```
