---
title: "02.5 Motif Processing"
output: html_document
date: '2025-21-01'
params: 
  metaChIPPath:
    label: "metaChIPPath"
    value: "metaChIP.tsv"
  refCoordsPath:
    label: "refCoordsPath"
    value: "refCoords.rds"
  motifDir: 
    label: "motifDir"
    value: ./motifData
  outDirMotifs: 
    label: "outDirMotifs"
    value: ./outDirMotifs
  outDirMeta:
    label: "outDirMeta"
    value: ./outDirMeta
---

```{r, setup}
suppressPackageStartupMessages({
  library(data.table)
  library(GenomicRanges)
  library(GenomeInfoDb)
})
```

```{r, paths}
motifDir <- params$motifDir
outDir <- params$outDirMotifs
outDirMeta <- params$outDirMeta

coordsPath <- params$refCoordsPath
chIPMetaPath <- params$metaChIPPath
```

## Data specification

### Motif Data

```{r, load meta data}
chIPMeta <- readRDS(chIPMetaPath)
allCoveredTfs <- unique(chIPMeta$tf_name)
allCoveredTfs <- allCoveredTfs[!is.na(allCoveredTfs)]
```

### Motif names

```{r}
motifFiles <- list.files(file.path(motifDir, "motifs"), full.names=TRUE)
motifFiles <- motifFiles[grepl(".rds", motifFiles)]
motifNames <- basename(motifFiles)
motifNames <- unlist(tstrsplit(motifNames, split=".", fixed=TRUE, keep=1))
```

### Reference Coordinates

```{r}
# DHS sites + mouse liftover DHS + ATLAS ATAC peaks
coords <- readRDS(coordsPath)
```

### Cofactors

```{r, get the cofactors}
library(OmnipathR)
library(tidyr)

  # get the cofactors
cofactors <- as.data.table(import_omnipath_interactions() %>% as_tibble())
# get tf-tf (prot) interactions
cofactors <- subset(cofactors, source_genesymbol %in% motifNames &   
                               target_genesymbol %in% motifNames)

# either source or target needs to be amongst the covered tfs
cofactors <- subset(cofactors, target_genesymbol %in% allCoveredTfs | 
                               source_genesymbol %in% allCoveredTfs)
allCofactors <- unique(c(cofactors$target_genesymbol, 
                         cofactors$source_genesymbol))
allTfs <- unique(c(allCofactors, allCoveredTfs))
saveRDS(allTfs, file.path(outDirMeta, "allTfs.rds"))
```

### Motif

TFs being profiled
All for which we have data + their cofactors
```{r, gathering motif names}
motifFiles <- list.files(file.path(motifDir, "motifs"), full.names=TRUE)
motifFiles <- motifFiles[grepl(".rds", motifFiles)]
motifNames <- basename(motifFiles)
motifNames <- unlist(tstrsplit(motifNames, split=".", fixed=TRUE, keep=1))
# subset to the ones either covered or being cofactors of covered
whichTfs <- motifNames %in% allTfs

allMotifNames <- motifNames[whichTfs]
motifFiles <- motifFiles[whichTfs]
motifNames <- motifNames[whichTfs]
```

```{r, preparing dimer motifs data, eval=TRUE}
dimerMotifs <- readRDS(file.path(motifDir, 
                                 "dimer_motifs",
                                 "dhs_dimer_motif_scoreMatrix.rds"))
outDirDimers <- file.path(outDir, "dimer_motifs")
if(!dir.exists(outDirDimers)) dir.create(outDirDimers)

for(dimer in colnames(dimerMotifs)){
  motif <- data.table(score=dimerMotifs[,dimer, drop=TRUE])
  motif <- cbind(as.data.table(coords), motif)
  motif <- makeGRangesFromDataFrame(as.data.frame(motif), keep.extra.columns=TRUE)
  
  saveRDS(motif, file.path(outDirDimers,
                           paste0(paste(dimer, "dimerMotif", sep="_"), ".rds")))
}
```

```{r, preparing dinuc motifs data, eval=TRUE}
dinucMotifs <- readRDS(file.path(motifDir, 
                                 "dinucleotide_motifs", 
                                 "dhs_dinuc_motif_scoreMatrix.rds"))
outDirDinuc <- file.path(outDir, "dinucleotide_motifs")
if(!dir.exists(outDirDinuc)) dir.create(outDirDinuc)

for(dinuc in colnames(dinucMotifs)){
  motif <- data.table(score=dinucMotifs[,dinuc, drop=TRUE])
  motif <- cbind(as.data.table(coords), motif)
  motif <- makeGRangesFromDataFrame(as.data.frame(motif), keep.extra.columns=TRUE)
  dinuc <- gsub("_", ":", dinuc)
  
  saveRDS(motif, file.path(outDirDinuc,
                           paste0(paste(dinuc, "dinucMotif", sep="_"), ".rds")))
}
```

```{r, sessioninfo}
sessionInfo()
```
