---
title: "02.7 Object construction"
output: html_document
date: '2025-21-01'
params: 
  motifDir: 
    label: "motifDir"
    value: ./motifs
---

```{r, setup}
suppressPackageStartupMessages({
library(BSgenome.Hsapiens.UCSC.hg38)
library(EnsDb.Hsapiens.v86)
library(parallel)
library(TFBlearner)
library(data.table)
library(GenomicRanges)
library(BiocParallel)
library(parallel)
})

seed <- 42
set.seed(seed)
```

```{r, params}
motifDir <- params$motifDir
```

## Prepare Data

### Meta data
```{r, prepare meta data}
# read all ATAC data & label test data
trainMetaAll <- readRDS("../data/meta_data/02_trainMetaAll.rds")
testMetaAll <- readRDS("../data/meta_data/02_testMetaAll.rds")

trainMetaChIP <- readRDS("../data/meta_data/03_trainMetaChIPAll.rds")
trainMetaChIP <- unique(trainMetaChIP, by="file_path")
trainMetaChIP[,full_id:=cellular_context_id]

testMetaChIP <- readRDS("../data/meta_data/03_testMetaChIPAll.rds")
testMetaChIP[,full_id:=cellular_context_id]

trainMetaAtac <- subset(trainMetaAll, !is.na(atac_path))
trainMetaAtac <- unique(trainMetaAtac, by="atac_path")
trainMetaAtac[,full_id:=context_id]
#trainMetaAtac <- subset(trainMetaAtac, context!="Glutamatergic_Neuron")

testMetaAtac <- subset(testMetaAll, !is.na(atac_path))
testMetaAtac <- unique(testMetaAtac, by="atac_path")
testMetaAtac[,full_id:=context_id]
```

### Reference Coordinates

```{r}
# DHS sites + mouse liftover DHS + ATLAS ATAC peaks
coords <- readRDS("../data/hybrid.custom.hg38.resized.resplit.rds")
```

### Cofactors

```{r, get the cofactors}
allTfs <- readRDS("../data/allTfs.rds")
allCofactors <- readRDS("../data/allCofactors.rds")
cofactorMap <- readRDS("../data/cofactorMaps.rds")
```

### Motif


```{r, attach additional motif data, eval=TRUE}
motifData <- list.files(motifDir, full.names=TRUE, recursive=TRUE)
motifData <- motifData[!(grepl("scoreMatrix", motifData))]
names(motifData) <- gsub(".rds","", basename(motifData))
```

## Tf promoter coordinates

```{r, tf promoter coordinates}
edb <- EnsDb.Hsapiens.v86
tfs <- c(allTfs,allCofactors)
txCoords <- transcripts(edb, 
                        columns=c("tx_seq_start"),
                        filter=GeneNameFilter(tfs))
end(txCoords) <- start(txCoords)+300
start(txCoords) <- start(txCoords)-300
txCoords <- as.data.table(txCoords)
txCoords[,seqnames:=paste0("chr", seqnames)]
setnames(txCoords, "gene_name", "tf_name")
txCoords <- makeGRangesFromDataFrame(as.data.frame(txCoords), 
                                     keep.extra.columns=TRUE)
names(txCoords) <- txCoords$tx_id
seqlevelsStyle(txCoords) <- "UCSC"
txCoords <- keepStandardChromosomes(txCoords, pruning.mode="coarse")
```

## Object construction

```{r arguments object, error=TRUE}
atacDataTrain <- c(trainMetaAtac$atac_path)
names(atacDataTrain) <- trainMetaAtac$full_id
atacDataTest <- c(testMetaAtac$atac_path)
names(atacDataTest) <- testMetaAtac$full_id
atacData <- as.list(c(atacDataTrain, atacDataTest))

chIPDataTrain <- c(trainMetaChIP$file_path)
names(chIPDataTrain) <- paste(trainMetaChIP$full_id, 
                              trainMetaChIP$tf_name, sep="_")
                              
chIPDataTest <- c(testMetaChIP$file_path)
names(chIPDataTest) <- paste(testMetaChIP$full_id,
                             testMetaChIP$tf_name, sep="_")
chIPData <- as.list(c(chIPDataTrain, chIPDataTest))
```

```{r small object, error=TRUE, eval=FALSE}
tfs <- c("JUN", "ATF3", "MYOG")
trainChIPMetaSmall <- subset(trainMetaChIP, tf_name %in% tfs)
trainAtacMetaSmall <- subset(trainMetaAtac, 
                             full_id %in% trainChIPMetaSmall$full_id)
atacDataSmall <- c(trainAtacMetaSmall$atac_path)
names(atacDataSmall) <- trainAtacMetaSmall$full_id

chIPDataSmall <- c(trainChIPMetaSmall$file_path)
names(chIPDataSmall) <- paste(trainChIPMetaSmall$full_id, 
                              trainChIPMetaSmall$tf_name, sep="_")

tfs <- c(tfs, subset(cofactorMap, tf_train %in% tfs)$tf_train_cofactor)
subNames <- grep(paste(tfs,collapse="|"), names(motifData), value=TRUE)
motifDataSmall <- motifData[names(motifData) %in% subNames] 

outDirSmall <- "./mae_small"
if(!dir.exists(outDirSmall)) dir.create(outDirSmall)
maeSmall <- prepData(refCoords=coords, 
                motifData=motifDataSmall,
                atacData=atacDataSmall,
                chIPData=chIPDataSmall,
                promoterCoords=txCoords,
                testSet=names(atacData)[1:2], 
                weightCol="rep_prob",
                isUncertainCol="is_uncertain", 
                saveHdf5=TRUE,
                outDir=outDirSmall)
saveRDS(maeSmall, file.path(outDirSmall, "mae_small_hdf5.rds"))
gc()
```

```{r, setup threads full, eval=TRUE}
BPPARAM <- SerialParam(force.GC=TRUE, log=TRUE, logdir=".")
```

```{r, construct object full, error=TRUE, eval=TRUE}
mae <- prepData(refCoords=coords, 
                motifData=motifData,
                atacData=atacData,
                chIPData=chIPData,
                promoterCoords=txCoords,
                testSet=names(atacDataTest), 
                weightCol="rep_prob",
                isUncertainCol="is_uncertain",
                saveHdf5=TRUE,
                outDir=".", 
                BPPARAM=BPPARAM)

#atacData <- append(atacData[names(atacData)!="98-2"],
#                   atacData[names(atacData)=="98-2"], after=0)

saveRDS(mae, "maeAllHdf5.rds")
gc()
```

```{r}
sessionInfo()
```
