---
title: "02.7 Object construction"
output: html_document
params: 
  motifDir: 
    label: "motifDir"
    value: ./motifs
  refCoordsPath:
    label: "refCoordsPath"
    value: "refCoords.rds"
  metaTrainPath:
    label: "metaTrainPath"
    value: "metaTrainPath.tsv"
  metaChIPTrainPath:
    label: "metaChIPTrainPath"
    value: "metaChIPTrainPath.tsv"
  metaTestPath:
    label: "metaTestPath"
    value: "metaTestPath.tsv"
  metaChIPTestPath:
    label: "metaChIPTestPath"
    value: "metaChIPTestPath.tsv"
  tfsListPath:
    label: "tfsListPath"
    value: "tfsListPath.tsv"
  cofactorMapPath:
    label: "cofactorMapPath"
    value: "cofactorMaps.tsv"
  metaDir:
    label: "metaDir"
    value: ./metaDir
  outDir:
    label: "outDir"
    value: ./outDir
---

```{r, setup}
suppressPackageStartupMessages({
library(BSgenome.Hsapiens.UCSC.hg38)
library(EnsDb.Hsapiens.v86)
library(parallel)
library(TFBlearner)
library(data.table)
library(GenomicRanges)
library(BiocParallel)
library(parallel)
})

seed <- 42
set.seed(seed)
```

```{r, params}
motifDir <- params$motifDir
outDir <- params$outDir
```

## Prepare Data

### Meta data
```{r, prepare meta data}
# read all ATAC data & label test data
trainMetaAll <- readRDS(params$metaTrainPath)
testMetaAll <- readRDS(params$metaTestPath)

trainMetaChIP <- readRDS(params$metaChIPTrainPath)
trainMetaChIP <- unique(trainMetaChIP, by="file_path")
trainMetaChIP[,full_id:=as.character(cellular_context_id)]
trainMetaChIP[,cellular_context_id:=as.integer(cellular_context_id)]
trainMetaChIP <- merge(trainMetaChIP, 
                       unique(trainMetaAll[,c("context_id", "context")], 
                              by="context_id"), 
                       by.x=c("cellular_context_id"),
                       by.y=c("context_id"), all.x=TRUE, all.y=FALSE)

testMetaChIP <- readRDS(params$metaChIPTestPath)
testMetaChIP[,full_id:=as.character(cellular_context_id)]
testMetaChIP[,cellular_context_id:=as.integer(cellular_context_id)]
testMetaChIP <- merge(testMetaChIP, 
                       unique(testMetaAll[,c("context_id", "context")], 
                              by="context_id"), 
                       by.x=c("cellular_context_id"),
                       by.y=c("context_id"), all.x=TRUE, all.y=FALSE)

trainMetaAtac <- subset(trainMetaAll, is_matched & !is.na(atac_path))
trainMetaAtac <- unique(trainMetaAtac, by="atac_path")
trainMetaAtac[,full_id:=as.character(context_id)]

testMetaAtac <- subset(testMetaAll, is_matched & !is.na(atac_path))
testMetaAtac <- unique(testMetaAtac, by="atac_path")
testMetaAtac[,full_id:=as.character(context_id)]

# TODO: Preppare ppred data in train and test!!!
predTrainMetaAtac <- subset(trainMetaAll, grepl("prediction", atac_path) | 
                                          context=="HUVEC_none")
predTrainMetaAtac$type <- "pred"
predTrainMetaAtac[,full_id:=as.character(context_id)]
predTrainMetaAtac <- subset(predTrainMetaAtac, !(full_id %in% trainMetaAtac$full_id))
predTrainMetaAtac <- subset(predTrainMetaAtac, grepl("_none", context))

predTestMetaAtac <- subset(testMetaAll, context %in% c("ESC_none", "Schwann_none"))
predTestMetaAtac$type <- "pred"
predTestMetaAtac[,full_id:=as.character(context_id)]
predTestMetaAtac <- subset(predTestMetaAtac, !(full_id %in% testMetaAtac$full_id))
predTestMetaAtac <- subset(predTestMetaAtac, grepl("_none", context))
# => make sure it is not in matched data! (well actually its okay just make sure its only train data, also in insertion profiles, double check the ones which arent from the preditiont directory!!
```

```{r, meta object}
trainMetaAtac$type <- "train"
trainMetaChIP$type <- "train"

# it is actually the same context
trainMetaAtac <- subset(trainMetaAtac, full_id!="33")
trainMetaAtac <- rbind(trainMetaAtac, predTrainMetaAtac)
trainMetaChIP[,full_id:=fifelse(full_id=="33", "165", full_id)]

testMetaAtac$type <- "test"
testMetaAtac <- rbind(testMetaAtac, predTestMetaAtac)
testMetaChIP$type <- "test"

metaTrainObject <- merge(trainMetaChIP[,c("full_id", "tf_name", "context", "file_path")], 
                         trainMetaAtac[,c("full_id","type", "atac_path", "context")], 
                         by.x=c("full_id"), 
                         by.y=c("full_id"), 
                         allow.cartesian=TRUE, all.x=TRUE, all.y=TRUE)
metaTrainObject[,type:=fifelse(is.na(type), "train", type)]
metaTrainObject[,is_matched:=fifelse(!is.na(atac_path) & !is.na(file_path), TRUE, FALSE)]
table(metaTrainObject$type, metaTrainObject$is_matched)
metaTrainObject$type <- "train"
metaTrainObject[,context:=fifelse(is.na(context.y), context.x, context.y)]
metaTrainObject$context.x <- metaTrainObject$context.y <- NULL

metaTestObject <- merge(testMetaChIP[,c("full_id","type", "tf_name", "context", "file_path")], 
                        testMetaAtac[,c("full_id","type", "context", "atac_path")], 
                        by.x=c("full_id", "type"), 
                        by.y=c("full_id", "type"), 
                        allow.cartesian=TRUE, all.x=TRUE, all.y=TRUE)
metaTestObject[,type:=fifelse(is.na(type), "test", type)]
metaTestObject[,is_matched:=fifelse(!is.na(atac_path) & !is.na(file_path), TRUE, FALSE)]
table(metaTestObject$type, metaTestObject$is_matched)
metaTestObject$type <- "test"
metaTestObject[,context:=fifelse(is.na(context.y), context.x, context.y)]
metaTestObject$context.x <- metaTestObject$context.y <- NULL

metaObject <- rbind(metaTrainObject, metaTestObject, fill=TRUE, use.names=TRUE)
metaObject[,context_name:=gsub("_none", "", context)]
saveRDS(metaObject, file.path(outDir, "01_maeAll_meta.rds"))

metaPred <- subset(metaObject, 
                   context_name %in% list.dirs("../data/atac/prediction", 
                                               full.names=FALSE))
metaPred <- unique(metaPred, by="full_id")
saveRDS(metaPred, file.path(outDir, "prediction_meta.rds"))
```


### Reference Coordinates

```{r}
# DHS sites + mouse liftover DHS + ATLAS ATAC peaks
coords <- readRDS(params$refCoordsPath)
```

### Cofactors

```{r, get the cofactors}
allTfs <- readRDS(params$tfsListPath)
cofactorMap <- fread(params$cofactorMapPath)

# this is not strictly needed (but has been used)
#allCofactors <- unique(c(cofactorMap$tf_train, cofactorMap$tf_train_cofactor))
```

### Motif


```{r, attach additional motif data, eval=TRUE}
motifData <- list.files(motifDir, full.names=TRUE, recursive=TRUE)
motifData <- motifData[!(grepl("scoreMatrix", motifData)) & 
                         grepl(".rds", motifData)]
names(motifData) <- gsub(".rds","", basename(motifData))
```

## Tf promoter coordinates

```{r, tf promoter coordinates}
edb <- EnsDb.Hsapiens.v86
tfs <- unique(allTfs) # unique(c(allTfs,allCofactors))
txCoords <- transcripts(edb, 
                        columns=c("tx_seq_start"),
                        filter=GeneNameFilter(tfs))
end(txCoords) <- start(txCoords)+300
start(txCoords) <- start(txCoords)-300
txCoords <- as.data.table(txCoords)
txCoords[,seqnames:=paste0("chr", seqnames)]
setnames(txCoords, "gene_name", "tf_name")
txCoords <- makeGRangesFromDataFrame(as.data.frame(txCoords), 
                                     keep.extra.columns=TRUE)
names(txCoords) <- txCoords$tx_id
seqlevelsStyle(txCoords) <- "UCSC"
txCoords <- keepStandardChromosomes(txCoords, pruning.mode="coarse")
```

## Object construction

```{r arguments object, error=TRUE}
atacDataTrain <- c(trainMetaAtac$atac_path)
names(atacDataTrain) <- trainMetaAtac$full_id
atacDataTest <- c(testMetaAtac$atac_path)
names(atacDataTest) <- testMetaAtac$full_id
atacData <- as.list(c(atacDataTrain, atacDataTest))

chIPDataTrain <- c(trainMetaChIP$file_path)
names(chIPDataTrain) <- paste(trainMetaChIP$full_id, 
                              trainMetaChIP$tf_name, sep="_")
                              
chIPDataTest <- c(testMetaChIP$file_path)
names(chIPDataTest) <- paste(testMetaChIP$full_id,
                             testMetaChIP$tf_name, sep="_")
chIPData <- as.list(c(chIPDataTrain, chIPDataTest))
```

```{r small object, error=TRUE, eval=FALSE}
tfs <- c("JUN", "ATF3", "MYOG")
trainChIPMetaSmall <- subset(trainMetaChIP, tf_name %in% tfs)
trainAtacMetaSmall <- subset(trainMetaAtac, 
                             full_id %in% trainChIPMetaSmall$full_id)
atacDataSmall <- c(trainAtacMetaSmall$atac_path)
names(atacDataSmall) <- trainAtacMetaSmall$full_id

chIPDataSmall <- c(trainChIPMetaSmall$file_path)
names(chIPDataSmall) <- paste(trainChIPMetaSmall$full_id, 
                              trainChIPMetaSmall$tf_name, sep="_")

tfs <- c(tfs, subset(cofactorMap, tf %in% tfs)$tf_cofactor)
subNames <- grep(paste(tfs,collapse="|"), names(motifData), value=TRUE)
motifDataSmall <- motifData[names(motifData) %in% subNames] 

outDirSmall <- file.path(outDir, "mae_small")
if(!dir.exists(outDirSmall)) dir.create(outDirSmall)
maeSmall <- prepData(refCoords=coords, 
                motifData=motifDataSmall,
                atacData=atacDataSmall,
                chIPData=chIPDataSmall,
                promoterCoords=txCoords,
                testSet=names(atacData)[1:2], 
                weightCol="rep_prob",
                isUncertainCol="is_uncertain", 
                saveHdf5=TRUE,
                outDir=outDirSmall)
saveRDS(maeSmall, file.path(outDirSmall, "mae_small_hdf5.rds"))
gc()
```

```{r, setup threads full, eval=TRUE}
BPPARAM <- SerialParam(force.GC=TRUE, log=TRUE, logdir=".")
```

```{r, construct object full, error=TRUE, eval=TRUE}
mae <- prepData(refCoords=coords, 
                motifData=motifData,
                atacData=atacData,
                chIPData=chIPData,
                promoterCoords=txCoords,
                testSet=names(atacDataTest), 
                weightCol="rep_prob",
                isUncertainCol="is_uncertain",
                saveHdf5=TRUE,
                outDir=outDir, 
                BPPARAM=BPPARAM)
saveRDS(mae, file.path(outDir, "01_maeAll.rds"))
gc()
```

```{r}
sessionInfo()
```
